code/modules/power/proto_sh_emitter.dm

```diff
@@ -0,0 +1,293 @@
+	if(default_deconstruction_crowbar(I))
+		return
+	return ..()
+
+/obj/machinery/power/proto_sh_emitter/attackby_secondary(obj/item/item, mob/user, params)
+	. = ..()
+	if(panel_open && item.tool_behaviour == TOOL_WRENCH)
+		if(default_unfasten_wrench(user, item))
+			return SECONDARY_ATTACK_CONTINUE_CHAIN
+
+/obj/machinery/power/proto_sh_emitter/process()
+	if(!is_on)
+		return
+	var/power_use = idle_power_usage
+	if(LAZYLEN(signs))
+		for(var/i in 1 to LAZYLEN(signs))
+			power_use += TILE_POWER_CONSUMPTION
+	use_power(power_use)
+
+/obj/machinery/power/proto_sh_emitter/proc/check_power()
+	if(is_operational)
+		return
+	var/turf/loc_turf = get_turf(src)
+	is_on = FALSE
+	has_barrier = FALSE
+	QDEL_LIST(signs)
+	update_icon_state()
+	message_admins("[src] turned off at [ADMIN_VERBOSEJMP(loc_turf)]")
+	log_game("[src] turned off at [AREACOORD(loc_turf)]")
+
+/** The vars you'll see in the proc() are referred to a mob looking north; NEx NEy refers to the North East corner x and y,
+ * all the other vars works in a similar way (N = North, S = South, E = East, W = West x = x axis, y = y axis, i = internal, o = outline). This way of naming the vars
+ * won't have much sense for the other directions, so always refer to the north direction when making changes as all other are already properly setup
+ *  This proc builds the barriers
+**/
+/obj/machinery/power/proto_sh_emitter/proc/build_barrier(SWxi,SWyi,NExi,NEyi,SWxo,SWyo,NExo,NEyo,mob/user)
+	to_chat(user, "<span class='warning'>You start to turn on the [src] and the generated shields!</span>")
+	if(!do_after(user, 1.5 SECONDS, target = src))
+		return
+	to_chat(user, "<span class='warning'>You turn on the [src] and the generated shields!</span>")
+	//Stores the outline of the room to generate
+	var/list/outline = list()
+	//Stores the internal turfs of the room to generate
+	var/list/internal = list()
+	var/turf/emitter_turf = get_turf(src)
+	message_admins("[src] turned on at [ADMIN_VERBOSEJMP(emitter_turf)] by [ADMIN_LOOKUPFLW(user)]")
+	log_game("[src] turned on at [AREACOORD(emitter_turf)] by [key_name(user)]")
+	is_on = TRUE
+	switch(dir) //this part check the direction of the machine and create the block in front of it
+		if(NORTH)
+			internal.Add(block(locate(x - SWxi, y + SWyi, z), locate(x + NExi, y + NEyi, z)))
+			outline.Add(block(locate(x - SWxo, y + SWyo, z), locate(x + NExo, y + NEyo, z)) - internal)
+		if(SOUTH)
+			internal.Add(block(locate(x - NExi, y - SWyi, z), locate(x + SWxi, y - NEyi, z)))
+			outline.Add(block(locate(x - NExo, y - SWyo, z), locate(x + SWxo, y - NEyo, z)) - internal)
+		if(EAST)
+			internal.Add(block(locate(x + SWyi, y - NExi, z), locate(x + NEyi, y + SWxi, z)))
+			outline.Add(block(locate(x + SWyo, y - NExo, z), locate(x + NEyo, y + SWxo, z)) - internal)
+		if(WEST)
+			internal.Add(block(locate(x - SWyi, y - SWxi, z), locate(x - NEyi, y + NExi, z)))
+			outline.Add(block(locate(x - SWyo, y - SWxo, z), locate(x - NEyo, y + NExo, z)) - internal)
```

Comment:
I really really do not like this. Maybe better var names will help?

Reply:
same but there isn't another way to do this, different var names will only increase the lenght of these and will make it incomprehensible